django/utils/datastructures.py

```diff
@@ -280,3 +282,94 @@ def __getitem__(self, key):
         if use_func:
             return self.func(value)
         return value
+
+
+def _destruct_iterable_mapping_values(data):
+    for i, elem in enumerate(data):
+        if len(elem) != 2:
+            raise ValueError("dictionary update sequence element #{} has "
+                             "length {}; 2 is required".format(i, len(elem)))
+
+        if not isinstance(elem[0], str):
+            raise ValueError('Element key invalid, only strings are allowed')
+
+        yield tuple(elem)
+
+
+def lowercased_key(method):
```

Comment:
I don't think this decorator is particularly needed. I think the code would be clearer if you just do the `key.lower()` inside `__getitem__`.